File: app/src/main/java/com/lemonade/dima/api/MockingInterceptor.java
package com.lemonade.dima.api;

import android.os.SystemClock;
import android.support.annotation.NonNull;

import java.security.SecureRandom;
import java.util.Random;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public class MockingInterceptor implements Interceptor {
    private final RequestsHandler mHandlers;
    private final Random mRandom;

    private MockingInterceptor() {
        mHandlers = new RequestsHandler();
        mRandom = new SecureRandom();
    }

    @NonNull
    public static Interceptor create() {
        return new MockingInterceptor();
    }

    @Override
    public Response intercept(Chain chain) {
        Request request = chain.request();
        Response response = mHandlers.proceed(request);
        int stubDelay = 500 + mRandom.nextInt(1500);
        SystemClock.sleep(stubDelay);
        return response;
    }
}

File: app/src/main/java/com/k4dima/chat/app/api/MockingInterceptor.kt
package com.k4dima.chat.app.api

import android.os.SystemClock
import com.k4dima.chat.messages.di.MessagesScope
import okhttp3.Interceptor
import java.security.SecureRandom
import java.util.concurrent.TimeUnit.SECONDS
import javax.inject.Inject

@MessagesScope
class MockingInterceptor
@Inject
internal constructor(private val handler: RequestsHandler) : Interceptor {
    override fun intercept(chain: Interceptor.Chain) =
            (SECONDS.toMillis(1) / 2)
                    .run { SystemClock.sleep(this + SecureRandom().nextInt(toInt())) }
                    .run { handler.proceed(chain.request()) }
}